### PR TITLE
Feature/qwen3 vl

### LINKED ISSUE #93: Qwen2.5-VL
## Model description
Qwen2.5-VL

## Open source examples / papers / references
* Paper: {https://arxiv.org/abs/2502.13923}
* Website: {https://chat.qwen.ai}
* Model code: {https://github.com/QwenLM/Qwen2.5-VL}
* Model weights: {https://huggingface.co/Qwen/Qwen2.5-VL-3B-Instruct}

## Why this model?
Qwen2.5-VL, the latest vision-language model, advances visual recognition, document parsing, and video comprehension through dynamic resolution processing, Window Attention, and a native Vision Transformer.

### PR DESCRIPTION
**Pull request by coder0143**
Resolves #93 

**Reference**
Ported Qwen3-VL model. Some things to be discussed @chapman20j @jenriver : 
1. Add support for FP8 format set of models (if possible).
2. Add sharding (mainly for 8B and 32B models).
3. Looking into bigger size image / video inputs. 
(vision features are variable, don't know if we have to manually setup jit compilation (using jax.lax api), also, jit compilation is quite expensive with vision inputs). 
4. Integrating into Sglang-jax. 

**Checklist**

- [x] I have read the **[Contribution Guidelines](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#contributing-a-model)** and used [pre-commit hooks](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#linting-and-type-checking) to format this commit.
- [x] I have added all the necessary **unit tests** for my change. (`run_model.py` for model usage, `test_outputs.py` and/or `model_validation_colab.ipynb` for quality).
- [x] **(If using an LLM)** I have carefully reviewed and removed all **superfluous comments** or unneeded, commented-out code. Only necessary and functional code remains.
- [x] I have signed the **[Contributor License Agreement (CLA)](https://cla.developers.google.com/about)**.

### CODE REVIEW

File: bonsai/models/qwen3_vl/modeling.py [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+    def __init__(self, dim: int, eps: float = 1e-6, *, rngs: nnx.Rngs):
+        self.eps = eps
+        self.weight = nnx.Param(jnp.ones((dim,), dtype=jnp.float32))
+
+    def __call__(self, x: Array) -> Array:
+        dtype = x.dtype
+        x_f32 = x.astype(jnp.float32)
+        rms = jax.lax.rsqrt(jnp.mean(x_f32**2, axis=-1, keepdims=True) + self.eps)
+        out = (x_f32 * rms) * self.weight[...]
+        return out.astype(dtype)
```
Comment (chapman20j) [inline bonsai/models/qwen3_vl/modeling.py]: Can remove `dtype=x.dtype` and just use `out.astype(x.dtype)` since `x` is never changed in this function.

File: bonsai/models/qwen3_vl/modeling.py [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+            use_bias=True,
+            rngs=rngs,
+        )
+
+    def __call__(self, hidden_states: Array) -> Array:
+        # Input: (num_patches, in_channels * temporal_patch_size * patch_size * patch_size)
+        cfg = self.config
+        seq_len = hidden_states.shape[0]
+
+        # Reshape to (1, D, H, W, C) for nnx.Conv (channels last)
```
Comment (chapman20j) [inline bonsai/models/qwen3_vl/modeling.py]: This comment doesn't correspond to the line of code following it

File: bonsai/models/qwen3_vl/modeling.py [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+            ),
+            text_config=Qwen3VLTextConfig(
+                hidden_size=2048,
+                intermediate_size=6144,
+                num_hidden_layers=28,
+                num_attention_heads=16,
+                num_key_value_heads=8,
+                tie_word_embeddings=True,
+            ),
+        )
```
[the remark this answers is not in the document]
Comment (chapman20j) [reply bonsai/models/qwen3_vl/modeling.py]: And similarly for the other methods in this config.

File: bonsai/models/qwen3_vl/modeling.py [hunk 4]
```diff
@@ -0,0 +1,1038 @@
...
+            use_bias=True,
+            kernel_sharding=self.shd_cfg.mlp_fc2_kernel,
+            rngs=rngs,
+        )
+
+    def __call__(self, x: Array) -> Array:
+        # Vision operates on (seq, hidden) without batch - no sharding benefit
+        x = self.linear_fc1(x, out_sharding=P(None, None))
+        x = nnx.gelu(x, approximate=True)
+        return self.linear_fc2(x, out_sharding=P(None, None))
```
Comment (chapman20j) [inline bonsai/models/qwen3_vl/modeling.py]: This will always remove the sharding on the outputs of the layers. Could you update this to have appropriate shardings?

**Review by jenriver — APPROVED** [review]
Thanks for this contribution, this is a really great addition!

One thing we could do later is to add jit to `forward_vision` after setting a default shape and padding, but we can leave it as a TODO for later.

**Comment by jenriver** [comment]
Hey! We just updated our contribution standards to keep the repo as lean as possible. Since this is already in great shape, could you do one quick final pass to squash your changes into a single commit and prune any extra comments or boilerplate?

We’re moving to the convention detailed here: https://github.com/jax-ml/bonsai/pull/159. Thanks for the help!

**Comment by chapman20j** [comment]
Closing since we merged in the other qwen3vl with fewer commits.
